File: snippets/visualbasic/System/AppDomain/DoCallBack/docallback_byref.vb
﻿Option Strict On
Option Explicit On

' <Snippet3>
Public Class PingPong
    Inherits MarshalByRefObject

    Private greetings As String = "PING!"
   
    Public Shared Sub Main()
        Dim otherDomain As AppDomain = AppDomain.CreateDomain("otherDomain")

        Dim pp As New PingPong()
        pp.MyCallBack()
        pp.greetings = "PONG!"
        otherDomain.DoCallBack(AddressOf pp.MyCallBack)

        ' Output:
        '   PING! from default domain
        '   PONG! from default domain
     End Sub

    ' Callback will always execute within defaultDomain due to inheritance from
    ' MarshalByRefObject
    Public Sub MyCallBack()
        Dim name As String = AppDomain.CurrentDomain.FriendlyName
        If name = AppDomain.CurrentDomain.SetupInformation.ApplicationName Then
            name = "defaultDomain"
        End If
        Console.WriteLine((greetings + " from " + name))
    End Sub

End Class
' </Snippet3>

File: snippets/fsharp/System/AppDomain/DoCallBack/docallback_byref.fs
module docallback_byref

// <Snippet3>
open System
type PingPong() as this =
    inherit MarshalByRefObject()
    [<DefaultValue>]
    val mutable greetings: string
    do  
        this.greetings <- "PING!"

    // Callback will always execute within defaultDomain due to inheritance from MarshalByRefObject
    member this.MyCallBack() =
        let mutable name = AppDomain.CurrentDomain.FriendlyName
        if name = AppDomain.CurrentDomain.SetupInformation.ApplicationName then
            name <- "defaultDomain"
        printfn $"{this.greetings} from {name}"

let otherDomain = AppDomain.CreateDomain "otherDomain"

let pp = new PingPong()
pp.MyCallBack()
pp.greetings <- "PONG!"
otherDomain.DoCallBack(CrossAppDomainDelegate pp.MyCallBack)

// Output:
//   PING! from defaultDomain
//   PONG! from defaultDomain
// </Snippet3>